File: solutions/257.Binary_Tree_Paths/AC_dfs_n.cpp
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_dfs_n.cpp
*  Create Date: 2015-08-16 09:59:42
*  Descripton:   
*/

#include <bits/stdc++.h>

using namespace std;
const int N = 0;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> ans;
        vector<int> path;
        dfs(root, path, ans);
        return ans;
    }

private:
    void dfs(TreeNode* root, vector<int> &path, vector<string> &ans) {
        if (!root) return;
        path.push_back(root->val);
        if (!root->left && !root->right)
            ans.push_back(getString(path));
        dfs(root->left, path, ans);
        dfs(root->right, path, ans);
        path.pop_back();
    }
    string getString(vector<int> path) {
        if (path.empty()) return "";
        string ret;
        for (int i = 0; i < path.size() - 1; ++i)
            ret += to_string(path[i]) + "->";
        ret += to_string(path[path.size() - 1]);
        return ret;
    }
};

int main() {

    return 0;
}

File: solutions/257.Binary_Tree_Paths/AC_dfs_n.py
#!/usr/local/bin/python
# -*- coding: utf-8 -*-
# Author:      illuz <iilluzen[at]gmail.com>
# File:        AC_dfs_n.py
# Create Date: 2015-08-16 10:15:54
# Usage:       AC_dfs_n.py 
# Descripton:  


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    # @param {TreeNode} root
    # @return {string[]}
    def binaryTreePaths(self, root):
        path =[]
        paths = []
        def dfs(root):
            if root:
                path.append(str(root.val))
                if root.left == None and root.right == None:
                    paths.append('->'.join(path))
                dfs(root.left)
                dfs(root.right)
                path.pop()
        dfs(root)
        return paths
            